File: state/ShippedOrderState.java
package state;

public class ShippedOrderState implements OrderState {
    private final OrderManagement orderManagement;

    public ShippedOrderState(OrderManagement orderManagement) {
        this.orderManagement = orderManagement;
    }

    @Override
    public void processOrder() {
        System.out.println("Cannot process the order at the current state");
    }

    @Override
    public void shipOrder() {
        System.out.println("The order is currently being shipped!");
    }

    @Override
    public void deliverOrder() {
        System.out.println("Delivering order now...");
        orderManagement.changeState(new DeliveredOrderState(orderManagement));
    }

    @Override
    public void cancelOrder() {
        System.out.println("Cannot cancel the order at the current state");
    }
}

File: typescript/patterns/state/shipped-order-state.ts
class ShippedOrderState implements OrderState {
  private orderManagement: OrderManagement;

  constructor(orderManagement: OrderManagement) {
    this.orderManagement = orderManagement;
  }

  processOrder(): void {
    console.log(
      `Cannot process the order: ${this.orderManagement
        .getOrder()
        .getName()} because it is already shipped.`,
    );
  }

  shipOrder(): void {
    console.log(
      `Order: ${this.orderManagement
        .getOrder()
        .getName()} is currently being shipping.`,
    );
  }

  deliverOrder(): void {
    console.log("Delivering order...");
    this.orderManagement.changeState(
      new DeliveredOrderState(this.orderManagement),
    );
  }

  cancelOrder(): void {
    console.log(
      `Cannot cancel the order: ${this.orderManagement
        .getOrder()
        .getName()} because it is already delivered.`,
    );
  }
}
